Makefile

```diff
@@ -14,7 +14,76 @@ build:
 	'$(VIRTUALENV_DIR)/bin/pip' install -r ./requirements.txt
 	'$(VIRTUALENV_DIR)/bin/python' setup.py install
 
+# Create a image with docker build & docker run.
+docker-build:
+	# e.g., docker build -t  <image_name> /path/to/Dockerfile/
+	# -t flag - to tag the image with a name so Docker recognizes this name.
+	docker build -t migrate2to3:v1 .
+
+docker-run: docker-build
+	# Runs docker-build part before docker-run part, i.e., docker-run DEPENDS on docker-build.
+	docker run -it migrate2to3:v1 bash
+
+# Create 2 virtualenvs. Test that Python 2.7 & Python 3.6 is running correctly
+# within each virtualenv respectively.
+# PYTHON 2.7 TESTS.
+test-help-tablesnap-py2.7: docker-build
+	docker run migrate2to3:v1 ./py27env/bin/python ./tablesnap --help
+
+test-help-tableslurp-py2.7: docker-build
+	docker run -e USER=root migrate2to3:v1 ./py27env/bin/python ./tableslurp --help
+
+test-help-tablechop-py2.7: docker-build
+	docker run migrate2to3:v1 ./py27env/bin/python ./tablechop --help
+
+test-help-tablesnap-sunset-py2.7: docker-build
+	docker run migrate2to3:v1 ./py27env/bin/python ./tablesnap-sunset --help
+
+test-help-tablesnap-sync-manifest-py2.7: docker-build
+	docker run migrate2to3:v1 ./py27env/bin/python ./tablesnap-sync-manifest --help
+
+# # PYTHON 3.6 TESTS.
+test-help-tablesnap-py3.6: docker-build
+	docker run migrate2to3:v1 ./py36env/bin/python ./tablesnap --help
+
+test-help-tableslurp-py3.6: docker-build
+	docker run -e USER=root migrate2to3:v1 ./py36env/bin/python ./tableslurp --help
+
+test-help-tablechop-py3.6: docker-build
+	docker run migrate2to3:v1 ./py36env/bin/python ./tablechop --help
+
+test-help-tablesnap-sunset-py3.6: docker-build
+	docker run migrate2to3:v1 ./py36env/bin/python ./tablesnap-sunset --help
+
+test-help-tablesnap-sync-manifest-py3.6: docker-build
+	docker run migrate2to3:v1 ./py36env/bin/python ./tablesnap-sync-manifest --help
+
+DIRECTORY=/tmp/tablesnap-test
+TIMEOUT_A=5
+TIMEOUT_B=10
+AWS_ACC_KEY_ID := $(shell aws configure get aws_access_key_id)
+AWS_SEC_ACC_KEY := $(shell aws configure get aws_secret_access_key)
+BUCKET=$(USER)-tablesnap-tests
+HARDLINKS_PATH=/tmp/tablesnap-hardlinks
+
+test-tablesnap-upload-py2.7: docker-build
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh python2 $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
+
+test-tablesnap-upload-py3.6: docker-build
+	docker run -it migrate2to3:v1 ./test_tablesnap.sh python3 $(DIRECTORY) $(TIMEOUT_A) $(TIMEOUT_B) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET) $(HARDLINKS_PATH)
+
+test-tableslurp-download-py2.7: docker-build
+	docker run -e USER=root -it migrate2to3:v1 ./test_tableslurp.sh python2 $(DIRECTORY) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET)
+
+test-tableslurp-download-py3.6: docker-build
+	docker run -e USER=root -it migrate2to3:v1 ./test_tableslurp.sh python3 $(DIRECTORY) $(AWS_ACC_KEY_ID) $(AWS_SEC_ACC_KEY) $(BUCKET)
```

Comment:
You should remove the `-it` on the `docker run` command lines in these four test targets too, just like you did with the `test-help-...` targets above. The `-it` isn't necessary and it will cause failures if you try to run `make` when standard input isn't a terminal.

For example, I tried to run all your targets that are named `test-...` by doing this:

```
grep '^test' Makefile | cut -f 1 -d: | xargs -n1 make
```

The `test-help-...` targets all work fine, but the `test-tablesnap-upload-...` and `test-tableslurp-download-...` targets fail because `docker run` complains that its standard input isn't a terminal.

Reply:
Deleted -it from lines that don't need user interaction.

Reply:
Looks good and I confirmed that I can now run all the tests with

```
grep '^test' Makefile | cut -f 1 -d: | xargs -n1 make
```